src/status.ts

```diff
@@ -0,0 +1,267 @@
+import * as vscode from "vscode";
+
+import {
+  isGemMissing,
+  addGem,
+  bundleCheck,
+  bundleInstall,
+  isGemOutdated,
+  updateGem,
+} from "./bundler";
+import { Ruby } from "./ruby";
+
+export const enum ServerCommand {
+  Start = "ruby-lsp.start",
+  Stop = "ruby-lsp.stop",
+  Error = "ruby-lsp.restart",
+}
+
+export class StatusItem {
+  private context: vscode.ExtensionContext;
+  private ruby: Ruby;
+  private selector: vscode.DocumentSelector;
+  private serverStatus: vscode.LanguageStatusItem;
+
+  constructor(context: vscode.ExtensionContext, ruby: Ruby) {
+    this.context = context;
+    this.ruby = ruby;
+    this.selector = {
+      scheme: "file",
+      language: "ruby",
+    };
+
+    this.serverStatus = this.createStatusItem(
+      "serverStatus",
+      "Ruby LSP: Starting...",
+      vscode.LanguageStatusSeverity?.Warning
+    );
+
+    this.serverStatus.command = {
+      title: "Restart server",
+      command: ServerCommand.Error,
+    };
+
+    this.createYjitStatus(this.ruby);
+    this.createRubyStatus(this.ruby);
+  }
+
+  public async updateStatus(status: ServerCommand) {
+    if (this.serverStatus) {
+      this.serverStatus.dispose();
+    }
+
+    switch (status) {
+      case ServerCommand.Start: {
+        this.serverStatus = this.createStatusItem(
+          "serverStatus",
+          "Ruby LSP: Running...",
+          vscode.LanguageStatusSeverity?.Warning
+        );
+
+        this.serverStatus.command = {
+          title: "Stop Server",
+          command: ServerCommand.Stop,
+        };
+
+        this.activateGemOutdatedButton();
+
+        break;
+      }
+      case ServerCommand.Stop: {
+        this.serverStatus = this.createStatusItem(
+          "serverStatus",
+          "Ruby LSP: Stopped",
+          vscode.LanguageStatusSeverity?.Warning
+        );
+
+        this.serverStatus.command = {
+          title: "Start Server",
+          command: ServerCommand.Start,
+        };
+
+        break;
+      }
+      case ServerCommand.Error: {
+        this.serverStatus = this.createStatusItem(
+          "serverStatus",
+          "Ruby LSP: Error",
+          vscode.LanguageStatusSeverity?.Error
+        );
+
+        this.serverStatus.command = {
+          title: "Restart Server",
+          command: ServerCommand.Error,
+        };
+      }
+    }
+  }
+
+  public async installGems(): Promise<boolean> {
+    if (await bundleCheck()) {
+      this.updateStatus(ServerCommand.Error);
+      const status: vscode.LanguageStatusItem = this.createStatusItem(
+        "installGems",
+        "Ruby LSP: The gems in the bundle are not installed.",
+        vscode.LanguageStatusSeverity?.Error
+      );
+
+      this.context.subscriptions.push(
+        vscode.commands.registerCommand("ruby-lsp.installGems", () => {
+          status.text = "Ruby LSP: Installing gems...";
+          status.command = undefined;
```

Comment:
I don't think you need to explicitly set it to `undefined`, that should be the default.